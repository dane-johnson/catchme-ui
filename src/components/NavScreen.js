import React from 'react'
import { Map, Marker, TileLayer, Circle } from 'react-leaflet'
import '../css/map.css'

const CENTER_RESET_DELAY = 6000

const getLatLng = (position) => [position.coords.latitude, position.coords.longitude]

class NavScreen extends React.Component{
  constructor(props) {
    super(props);
    const { position } = props;
    this.state = {
      center: getLatLng(position),
      followMarker: true
    }
    this.handleMoveend = this.handleMoveend.bind(this);
    this.returnToMarker = this.returnToMarker.bind(this);
    this.handleMove = this.handleMove.bind(this);
  }
  handleMove() {
    const { timeoutId } = this.state
    if (!timeoutId) return
    window.clearTimeout(timeoutId)
    this.setState({timeoutId: null})
  }
  handleMoveend(ev) {
    const center = ev.target.getCenter()
    const timeoutId = window.setTimeout(this.returnToMarker, CENTER_RESET_DELAY)
    this.setState({center, timeoutId, followMarker: false})
  }
  returnToMarker() {
    const { position } = this.props
    this.setState({center: getLatLng(position), followMarker: true})
  }
  render() {
    const { position, size } = this.props;
    const { center, followMarker } = this.state;
    if (!position) return null;
    const me = getLatLng(position);
    const mapCenter = (followMarker ? me : center)
    return (
      <Map
        center={mapCenter}
        zoom={20}
        onMove={this.handleMove}
        onMoveend={this.handleMoveend}
        useFlyTo={true}
        >
        <TileLayer
          url='http://{s}.tile.osm.org/{z}/{x}/{y}.png'
          attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
        />
        <Marker position={me}/>
        <Circle center={me} radius={size}/>
      </Map>
    )
  }
}

export default NavScreen
