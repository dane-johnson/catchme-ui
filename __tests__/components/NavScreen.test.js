import React from 'react'
import renderer from 'react-test-renderer'
import NavScreen from '../../src/components/NavScreen'

jest.mock("react-leaflet")

const mockPosition = {
  coords: {
    latitude: 6.9,
    longitude: 9.6,
  }
}

describe('<NavScreen />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <NavScreen position={mockPosition} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
