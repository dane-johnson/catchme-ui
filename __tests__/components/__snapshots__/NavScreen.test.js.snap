exports[`<NavScreen /> renders correctly 1`] = `null`;
